pkgs/applications/editors/eclipse/plugins.nix

```diff
@@ -15,80 +34,123 @@ rec {
+                  <source>
+                    ${if repository != null then "<repository location=\"${repository}\" />" else ""}
+                    ${if metadataRepository != null then "<repository location=\"${metadataRepository}\" kind=\"metadata\" />" else ""}
+                    ${if artifactRepository != null then "<repository location=\"${artifactRepository}\" kind=\"artifact\" />" else ""}
+                  </source>
+                  <destination compressed="true" name="${name}" location="file:'$out'" />
+                  ${iuEntries}
+                  <slicingOptions followStrict="true"/>
+                </p2.mirror>
+              </target>
+            </project>' > /build/build.xml
+          ${eclipse}/bin/eclipse \
+            -noSplash \
+            -configuration /build/configuration \
+            -application org.eclipse.ant.core.antRunner \
+            -buildfile /build/build.xml
+        '' + (sanitizeUpdateSite "$out");
+
+        nativeBuildInputs = [ zip eclipse xmlstarlet ];
+        outputHash = hash;
+        outputHashMode = "recursive";
+
+        passthru = {
+          installableUnits = map (iu: "${iu.id}/${iu.version}") ius;
+        };
+
+        inherit repository artifactRepository metadataRepository;
+      };
+
+  # Publish the given plugin and feature jars as a p2 update site.
+  buildEclipsePlugin = { name
+                       , srcFeature ? null
+                       , srcFeatures ? []
+                       , srcPlugin ? null
+                       , srcPlugins ? []
+                       , eclipse ? eclipses.eclipse-sdk
+                       , ... } @ attrs:
+    assert (srcPlugin == null) != (srcPlugins == []);
+    assert (srcFeature != null) -> (srcFeatures == []);
+
+    let
+      fSrcs = if (srcFeature != null) then [ srcFeature ] else srcFeatures;
+      pSrcs = if (srcPlugin != null) then [ srcPlugin ] else srcPlugins;
+
+    in
+      buildEclipsePluginBase (attrs // {
+        srcs = fSrcs ++ pSrcs;
+
+        buildCommand = ''
+          mkdir features
+          for feature in ${toString fSrcs}; do
+            cp -v $feature features/$(stripHash $feature)
+          done
+
+          mkdir plugins
+          for plugin in ${toString pSrcs}; do
+            cp -v $plugin plugins/$(stripHash $plugin)
+          done
+
+          ${eclipse}/bin/eclipse \
+            -noSplash \
+            -configuration /build/configuration \
+            -application org.eclipse.equinox.p2.publisher.FeaturesAndBundlesPublisher \
+            -metadataRepository file:$out \
+            -artifactRepository file:$out \
+            -publishArtifacts \
+            -source . \
+            -compress
+        '';
 
-            mkdir -p $dropinDir/plugins
-            for plugin in ${toString pSrcs}; do
-              cp -v $plugin $dropinDir/plugins/$(stripHash $plugin)
-            done
-          '';
-        });
+        nativeBuildInputs = [ eclipse ];
+      });
```

Comment:
This shouldn't be used under normal circumstances, but it needs to be here for two reasons.
- Backwards compatibility.
- Plugins which aren't published in p2 repos. This is *extremely* rare these days in the Eclipse ecosystem!